### PR TITLE
fix: seek_to on Kafka external stream could cause core dump

### LINKED ISSUE #285: use seek_to='earliest' with external stream will core dump.
```sql
-- redpanda
rpk topic create test

-- proton
drop stream if exists test_external1;
create external stream if not exists test_external1(id int, value string) settings type = 'kafka', brokers = 'redpanda-1:9092', topic = 'test', data_format='JSONEachRow';

-- redpanda
rpk topic produce test
{"id":1,"value":"sometext"}
Produced to partition 0 at offset 2 with timestamp 1700016423153.
{"id":2,"value":"aaaaa"}
Produced to partition 0 at offset 3 with timestamp 1700016428203.
{"id":3,"value":"bbbbb"}
Produced to partition 0 at offset 4 with timestamp 1700016431238.
{"id":4,"value":"ccccc"}
Produced to partition 0 at offset 5 with timestamp 1700016434389.

-- proton
select * from test_external1 settings seek_to='earliest';
-- no results and Core Dump

```

```
2023.11.16 13:19:20.279404 [ 304985 ] {3feeacc9-6851-493d-bdba-4efae38d3895} <Fatal> : Logical error: 'Bad cast from type DB::ColumnVector<int> to DB::ColumnString'.
2023.11.16 13:19:20.280532 [ 305421 ] {} <Fatal> BaseDaemon: ########################################
2023.11.16 13:19:20.280962 [ 305421 ] {} <Fatal> BaseDaemon: (version 1.3.22, build id: 5613AA59E0EF211617341403B5ED5344D382662E) (from thread 304985) (query_id: 3feeacc9-6851-493d-bdba-4efae38d3895) (query: select * from test_external1 settings seek_to='earliest';) Received signal Aborted (6)
2023.11.16 13:19:20.281244 [ 305421 ] {} <Fatal> BaseDaemon: 
2023.11.16 13:19:20.281797 [ 305421 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f7ff838300b 0x7f7ff8362859 0x34153d9b 0x34153f1d 0x34154153 0x1e8e2e2c 0x489adb2a 0x595471da 0x59546b02 0x59546aad 0x596615b6 0x59551c8d 0x595460b1 0x4a81bfe9 0x4a81b4f3 0x4a925b37 0x4a9255cf 0x4a8ae5a3 0x4a8af046 0x4a8b4f7f 0x4a8b4e15 0x4a8b4df9 0x4a8b4d3d 0x4a8b4ad4 0x4a8b4775 0x4a8b4755 0x4a8b4735 0x4a8b4703 0x34213343 0x34208ab5 0x34424115 0x34436225 0x344360d5 0x34435f9d 0x34435656 0x7f7ff854a609 0x7f7ff845f133
2023.11.16 13:19:20.282169 [ 305421 ] {} <Fatal> BaseDaemon: 4. raise @ 0x7f7ff838300b in ?
2023.11.16 13:19:20.282451 [ 305421 ] {} <Fatal> BaseDaemon: 5. abort @ 0x7f7ff8362859 in ?
2023.11.16 13:19:20.386376 [ 305421 ] {} <Fatal> BaseDaemon: 6. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Common/Exception.cpp:0: DB::abortOnFailedAssertion(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) @ 0x34153d9b in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:20.454533 [ 305421 ] {} <Fatal> BaseDaemon: 7. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Common/Exception.cpp:63: DB::handle_error_code(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, bool, std::__1::vector<void*, std::__1::allocator<void*>> const&) @ 0x34153f1d in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:20.521590 [ 305421 ] {} <Fatal> BaseDaemon: 8. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Common/Exception.cpp:70: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, bool) @ 0x34154153 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:21.000602 [ 304943 ] {} <Information> AsynchronousMetrics: MemoryTracking: was 1.84 GiB, peak 1.84 GiB, free memory in arenas 0.00 B, will set to 1.90 GiB (RSS), difference: 65.02 MiB
2023.11.16 13:19:21.170451 [ 305421 ] {} <Fatal> BaseDaemon: 9. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Common/assert_cast.h:47: DB::ColumnString const& assert_cast<DB::ColumnString const&, DB::IColumn const&>(DB::IColumn const&) @ 0x1e8e2e2c in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:21.266746 [ 305421 ] {} <Fatal> BaseDaemon: 10. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Columns/ColumnString.cpp:118: DB::ColumnString::insertRangeFrom(DB::IColumn const&, unsigned long, unsigned long) @ 0x489adb2a in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:21.452815 [ 305421 ] {} <Fatal> BaseDaemon: 11. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Storages/ExternalStream/Kafka/KafkaSource.cpp:194: DB::KafkaSource::parseFormat(rd_kafka_message_s const*) @ 0x595471da in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:21.606242 [ 305421 ] {} <Fatal> BaseDaemon: 12. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Storages/ExternalStream/Kafka/KafkaSource.cpp:125: DB::KafkaSource::doParseMessage(rd_kafka_message_s const*, unsigned long) @ 0x59546b02 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:21.760038 [ 305421 ] {} <Fatal> BaseDaemon: 13. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Storages/ExternalStream/Kafka/KafkaSource.cpp:120: DB::KafkaSource::parseMessage(void*, unsigned long, void*) @ 0x59546aad in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:21.883039 [ 305421 ] {} <Fatal> BaseDaemon: 14. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/KafkaLog/KafkaWALSimpleConsumer.cpp:450: klog::KafkaWALSimpleConsumer::consume(void (*)(void*, unsigned long, void*), void*, unsigned int, int, klog::KafkaWALContext const&) const @ 0x596615b6 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.000479 [ 304943 ] {} <Information> AsynchronousMetrics: MemoryTracking: was 1.90 GiB, peak 1.90 GiB, free memory in arenas 0.00 B, will set to 2.04 GiB (RSS), difference: 144.41 MiB
2023.11.16 13:19:22.000738 [ 304943 ] {} <Debug> MemoryTracker: Current memory usage (total): 2.04 GiB.
2023.11.16 13:19:22.021055 [ 305421 ] {} <Fatal> BaseDaemon: 15. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Storages/ExternalStream/Kafka/KafkaSource.cpp:103: DB::KafkaSource::readAndProcess() @ 0x59551c8d in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.162947 [ 305421 ] {} <Fatal> BaseDaemon: 16. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Storages/ExternalStream/Kafka/KafkaSource.cpp:81: DB::KafkaSource::generate() @ 0x595460b1 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.243379 [ 305421 ] {} <Fatal> BaseDaemon: 17. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Processors/ISource.cpp:138: DB::ISource::tryGenerate() @ 0x4a81bfe9 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.326628 [ 305421 ] {} <Fatal> BaseDaemon: 18. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Processors/ISource.cpp:103: DB::ISource::work() @ 0x4a81b4f3 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.391552 [ 305421 ] {} <Fatal> BaseDaemon: 19. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Processors/Executors/ExecutionThreadContext.cpp:46: DB::executeJob(DB::ExecutingGraph::Node*, DB::ReadProgressCallback*) @ 0x4a925b37 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.458116 [ 305421 ] {} <Fatal> BaseDaemon: 20. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Processors/Executors/ExecutionThreadContext.cpp:84: DB::ExecutionThreadContext::executeTask() @ 0x4a9255cf in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.560971 [ 305421 ] {} <Fatal> BaseDaemon: 21. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Processors/Executors/PipelineExecutor.cpp:271: DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x4a8ae5a3 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.660475 [ 305421 ] {} <Fatal> BaseDaemon: 22. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Processors/Executors/PipelineExecutor.cpp:236: DB::PipelineExecutor::executeSingleThread(unsigned long) @ 0x4a8af046 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.761690 [ 305421 ] {} <Fatal> BaseDaemon: 23. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Processors/Executors/PipelineExecutor.cpp:373: DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0::operator()() const @ 0x4a8b4f7f in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.866410 [ 305421 ] {} <Fatal> BaseDaemon: 24. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&) @ 0x4a8b4e15 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:22.976268 [ 305421 ] {} <Fatal> BaseDaemon: 25. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/tuple:1789: decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&, std::__1::tuple<>&>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&, std::__1::tuple<>&, std::__1::__tuple_indices<>) @ 0x4a8b4df9 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.078951 [ 305421 ] {} <Fatal> BaseDaemon: 26. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/tuple:1798: decltype(auto) std::__1::apply[abi:v15000]<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&, std::__1::tuple<>&>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&, std::__1::tuple<>&) @ 0x4a8b4d3d in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.176044 [ 305421 ] {} <Fatal> BaseDaemon: 27. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Common/ThreadPool.h:202: ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&&)::'lambda'()::operator()() @ 0x4a8b4ad4 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.278742 [ 305421 ] {} <Fatal> BaseDaemon: 28. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&&)::'lambda'()&>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&&) @ 0x4a8b4775 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.381950 [ 305421 ] {} <Fatal> BaseDaemon: 29. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/__functional/invoke.h:480: void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&&)::'lambda'()&>(ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&&)::'lambda'()&) @ 0x4a8b4755 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.452575 [ 304754 ] {} <Debug> DNSResolver: Updating DNS cache
2023.11.16 13:19:23.452707 [ 304754 ] {} <Debug> DNSResolver: Updated DNS cache
2023.11.16 13:19:23.484039 [ 305421 ] {} <Fatal> BaseDaemon: 30. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/__functional/function.h:235: std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&&)::'lambda'(), void ()>::operator()[abi:v15000]() @ 0x4a8b4735 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.588186 [ 305421 ] {} <Fatal> BaseDaemon: 31. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/__functional/function.h:717: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0>(DB::PipelineExecutor::executeImpl(unsigned long, DB::ExecuteMode)::$_0&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) @ 0x4a8b4703 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.631865 [ 305421 ] {} <Fatal> BaseDaemon: 32. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/__functional/function.h:848: std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const @ 0x34213343 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.677863 [ 305421 ] {} <Fatal> BaseDaemon: 33. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/__functional/function.h:1187: std::__1::function<void ()>::operator()() const @ 0x34208ab5 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.727970 [ 305421 ] {} <Fatal> BaseDaemon: 34. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Common/ThreadPool.cpp:295: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x34424115 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.781663 [ 305421 ] {} <Fatal> BaseDaemon: 35. /home/timeplus/lijianan/ssd/gitdownload/proton_/src/Common/ThreadPool.cpp:140: void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const @ 0x34436225 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.839696 [ 305421 ] {} <Fatal> BaseDaemon: 36. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) @ 0x344360d5 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.900407 [ 305421 ] {} <Fatal> BaseDaemon: 37. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/thread:285: void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) @ 0x34435f9d in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.954731 [ 305421 ] {} <Fatal> BaseDaemon: 38. /home/timeplus/lijianan/ssd/gitdownload/proton_/contrib/llvm-project/libcxx/include/thread:295: void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x34435656 in /home/timeplus/lijianan/ssd/gitdownload/proton_/build/programs/proton
2023.11.16 13:19:23.954906 [ 305421 ] {} <Fatal> BaseDaemon: 39. ? @ 0x7f7ff854a609 in ?
2023.11.16 13:19:23.954971 [ 305421 ] {} <Fatal> BaseDaemon: 40. clone @ 0x7f7ff845f133 in ?
2023.11.16 13:19:23.955027 [ 305421 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2023.11.16 13:19:27.000342 [ 304943 ] {} <Information> AsynchronousMetrics: MemoryTracking: was 2.14 GiB, peak 2.14 GiB, free memory in arenas 0.00 B, will set to 2.06 GiB (RSS), difference: -80.25 MiB
```

### PR DESCRIPTION
**Pull request by zliang-min**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:

closes #285

### CODE REVIEW

File: src/Storages/ExternalStream/Kafka/KafkaSource.cpp [hunk 1]
```diff
@@ -190,8 +190,8 @@ void KafkaSource::parseFormat(const rd_kafka_message_t * kmessage)
         {
             /// Merge all data in the current batch into the same chunk to avoid too many small chunks
             auto new_data(format_executor->getResultColumns());
-            for (size_t pos = 0; auto & column : current_batch)
-                column->insertRangeFrom(*new_data[pos], 0, new_rows);
+            for (size_t pos = 0; pos < current_batch.size(); ++pos)
+                current_batch[pos]->insertRangeFrom(*new_data[pos], 0, new_rows);
```
Comment (yl-lisen) [inline src/Storages/ExternalStream/Kafka/KafkaSource.cpp:194]: looks the another path `slower path` has same problem.

**Review by yl-lisen — APPROVED** [review]
LGTM

**Comment by jovezhong** [comment]
(Jove Github Bot) added it to the current sprint.

**Comment by jovezhong** [comment]
(Jove Github Bot) moved this ticket out of the GitHub project(up to 1200 tickets for one project).

**Pull request merged**